paseos/tests/communication_test.py

```diff
@@ -0,0 +1,611 @@
+"""Tests to check the communication models creation and calculation function(s)"""
+import numpy as np
+import pykep as pk
+import pytest
+
+from paseos.actors.actor_builder import ActorBuilder
+from paseos.actors.ground_station_actor import GroundstationActor
+from paseos.actors.spacecraft_actor import SpacecraftActor
+from paseos.communication.device_type import DeviceType
+from paseos.communication.optical_link_model import OpticalLinkModel
+from paseos.communication.optical_receiver_model import OpticalReceiverModel
+from paseos.communication.optical_transmitter_model import OpticalTransmitterModel
+from paseos.communication.radio_link_model import RadioLinkModel
+from paseos.communication.radio_receiver_model import RadioReceiverModel
+from paseos.communication.radio_transmitter_model import RadioTransmitterModel
+from paseos.communication.receiver_model import ReceiverModel
+from paseos.communication.transmitter_model import TransmitterModel
+
+
+def test_link_creation():
+    t0 = pk.epoch_from_string("2023-Jan-04 20:00:00")
+    maspalomas_groundstation = ActorBuilder.get_actor_scaffold(
+        name="maspalomas_groundstation", actor_type=GroundstationActor, epoch=t0
+    )
+    receiver_name = "maspalomas_radio_receiver_1"
+
+    ActorBuilder.add_comm_device(
+        actor=maspalomas_groundstation,
+        device_name=receiver_name,
+        noise_temperature=135,
+        line_losses=1,
+        polarization_losses=3,
+        antenna_gain=62.6,
+        device_type=DeviceType.RADIO_RECEIVER,
+    )
+
+    sat_actor: SpacecraftActor = ActorBuilder.get_actor_scaffold(
+        name="Sat", actor_type=SpacecraftActor, epoch=t0
+    )
+    radio_name = "sat_radio_transmitter_1"
+    optical_transmitter_name = "sat_optical_transmitter_1"
+
+    ActorBuilder.add_comm_device(
+        actor=sat_actor,
+        device_name=optical_transmitter_name,
+        input_power=1,
+        power_efficiency=1,
+        antenna_efficiency=1,
+        line_losses=1,
+        point_losses=3,
+        fwhm=1e-3,
+        device_type=DeviceType.OPTICAL_TRANSMITTER,
+    )
+
+    ActorBuilder.add_comm_device(
+        actor=sat_actor,
+        device_name=radio_name,
+        input_power=2,
+        power_efficiency=0.5,
+        antenna_efficiency=0.5,
+        line_losses=1,
+        point_losses=5,
+        antenna_diameter=0.3,
+        device_type=DeviceType.RADIO_TRANSMITTER,
+    )
+
+    comms_sat: SpacecraftActor = ActorBuilder.get_actor_scaffold(
+        name="Comms", actor_type=SpacecraftActor, epoch=t0
+    )
+    optical_receiver_name = "optical_receiver_1"
+    ActorBuilder.add_comm_device(
+        actor=comms_sat,
+        device_name=optical_receiver_name,
+        line_losses=4.1,
+        antenna_gain=114.2,
+        device_type=DeviceType.OPTICAL_RECEIVER,
+    )
+
+    with pytest.raises(
+            AssertionError,
+            match="An optical transmitter is required for this optical link.",
+    ):
```

Comment:
Why are these raised everywhere? This seems to indicate a problem in the structure to me if by default your tests raise exception that you have to capture

Reply:
A number of these specic assertions are now removed because of imrpoved code structure. The rest are testing the asserts in ActorBuilder. Checking that are errors are raised when parameters such as input power are set as negative.